### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized perc/cbst model with integer arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#321]
Public recipe.train().optimizer(adamw).lr(0.001).seed(19858644896192).epochs(1).log(all).stop(0.0).f(6,9).run() on backend:nv0 stalls at 66s before any epoch arithmetic. Earliest public failure is recipe.train().run() at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896). The reproduction uses data("data/ordinal/ordered_class_files").target("target").split(0.8), loaded but ignored at the stall point; the model is .perc(8).selu().norm(batch) then .cbst().silu().qi(3).k.s with .loss(bce). Perc lowers via lower_project (recipe.rs 5113) and cbst as Estimator (recipe.rs 3074) with quantized .qi(3).k.s nodes; this perc/cbst quantized path is not represented by 317 (pool/cbst), 315 (perc/MoE), or 302 (perc/LSTM). Small ordinal load and few operations justify sub-second arithmetic, so the 66s gap is native artifact compilation, not throughput.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `f781466a7fbd31eb`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized perc/cbst model with integer arithmetic",
  "body": "Public recipe.train().optimizer(adamw).lr(0.001).seed(19858644896192).epochs(1).log(all).stop(0.0).f(6,9).run() on backend:nv0 stalls at 66s before any epoch arithmetic. Earliest public failure is recipe.train().run() at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896). The reproduction uses data(\"data/ordinal/ordered_class_files\").target(\"target\").split(0.8), loaded but ignored at the stall point; the model is .perc(8).selu().norm(batch) then .cbst().silu().qi(3).k.s with .loss(bce). Perc lowers via lower_project (recipe.rs 5113) and cbst as Estimator (recipe.rs 3074) with quantized .qi(3).k.s nodes; this perc/cbst quantized path is not represented by 317 (pool/cbst), 315 (perc/MoE), or 302 (perc/LSTM). Small ordinal load and few operations justify sub-second arithmetic, so the 66s gap is native artifact compilation, not throughput.",
  "rationale": "Performance packet, backend nv0, stable:true, 66s > 60s, quantized model nodes trigger compile_native_artifact before epoch work. Inspected source shows perc and cbst have distinct lowering paths from pool/cbst (317) and perc/MoE (315), and no existing nv0 issue represents this earliest cause. Loader, seed, loss, and quantization-format differences are at/after the failure point and do not change the implementation path, but the perc + cbst combination materially expands the unreported nv0 quantized native-compilation stall, requiring a new issue."
}
```

## Deterministic failure packet

kind=performance
id=f781466a7fbd31eb
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10048 next:10049 composition:19858644896192
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 66 seconds
measurement=elapsed_seconds:66 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10048-309.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/ordinal/ordered_class_files")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.perc(8).selu().norm(batch)
		.cbst().silu().qi(3).k.s
		.loss(bce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(19858644896192)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #429.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #429: Verify current NVIDIA Perc and CatBoost execution</summary>

Closes #321

This is a verification-only PR. It intentionally changes no tracked files.

## Current public result

Commit: `9098de8711552004bb9eabd84c6d8d9c120d4c58`

Command:

```text
RECIPE_DEBUG=1 ./target/release/recipe --device benji:nv0 /home/nate/.cache/recipe/issues/321/reproduction.rs
```

The exact issue data path, split, Perc/CatBoost model, training settings, and custom `f(6,9)` arithmetic complete successfully from cold artifacts:

- Four issue-specific artifacts begin with `cache=miss`: the prefix, CatBoost surrogate, 16,000-row training, and 4,000-row validation artifacts.
- Clang emits NVIDIA LLVM bitcode, and `llc` emits PTX for each cold artifact.
- The training epoch completes in 340.514 ms.
- Training BCE is finite at 0.7064, and held-out BCE is finite at 0.6989.
- The full training and evaluation process exits 0 after 17.03 seconds.

The reported 66-second NVIDIA native-artifact stall does not reproduce on current `minimal`. No production-code change is supported by this result.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
